### DISCUSSION TITLE
algorithms/ads/4

### DISCUSSION THREAD

**Discussion by giscus** [NoughtQ/notebook#16] [Announcements]
# algorithms/ads/4

NoughtQ的笔记本

https://note.noughtq.top/algorithms/ads/4

**Comment by fang88666** [comment]
“再根据左偏树的定义知，左子树的节点数不小于右子树的节点数”这句话似乎不太严谨？这里好像就给出了一个反例https://note.isshikih.top/cour_note/D2CX_AdvancedDataStructure/Lec04/#%E9%80%92%E5%BD%92%E5%BC%8F

**Reply by NoughtQ** [reply]
啊哦，这里确实写的有问题，应该表述为“**左孩子的空指针路径长度不小于右孩子的空指针路径长度**”，之后上传的时候我会修改一下，感谢指正 :smile:

**Comment by XuebaStudy** [comment]
"所谓「重节点」，指的是满足右子树的节点个数占整棵树节点个数（不包括这个根节点）的一半及以上的节点。"应该是要包含根节点？否则“例子“中节点3应该也为heavy node，但其实并不。

**Reply by NoughtQ** [reply]
这里是因为我在看cy的PPT时眼花了：我把上一行的note看成了not，跟下面的include混在一起看作not include了 :dizzy_face:。这个问题我早上复习期中考的时候也发现这个问题了，但如果没有你的提醒，我都忘记要改过来了，万分感谢 :pray:

**Comment by XuebaStudy** [comment]
"由这个定理，不难得到以下推论：对于一个具有 N 个节点的左偏堆，它的右子树上至多有 ⌊log(N+1)⌋ 个节点。"
但是如果以下两个左式堆合并后的左式堆是不是就不符了。
  0              
/
1

   2
/
 3
/    \   
 4     5
/  \  /  \
6 7 8  9

**Reply by NoughtQ** [reply]
对于你举出的例子，我的意见是：
1. 这张图最底下漏画了连接9的那条边，那我就假设9是5的右孩子了
2. 这两个堆合并后很明显右子树很小呀，根节点0只有1个右孩子1，你是不是合并完忘记交换子树了呢？

**Reply by XuebaStudy** [reply]
1. 我确实是“9是5的右孩子”的意思，这个评论giscus不知道为什么没显示\ ......
2. 合并前1是0的左孩子，2大于0，所以以2为根的堆应与0的右孩子合并（即直接成为0的右孩子）：合并后：
第一层：0
第二层：1，2（0的左右孩子）
第三层：3（2的左孩子）
第四层：4，5（3的左右孩子）
......
这样一步到位，符合左偏堆性质，不用交换吧（话说giscus有没有方便演示图片的方式捏）

**Reply by NoughtQ** [reply]
不知是不是你的笔误，5应该是2的右孩子吧。我列出了每个节点对应的Npl值：
|节点|深度|Npl|
|:-|:-|:-|
|0|0|1|
|1|1|0|
|2|1|1|
|3|2|0|
|4|3|1|
|5|2|1|
|6|4|0|
|7|4|0|
|8|3|0|
|9|3|0|

因此0的右孩子2的Npl值(1)大于左孩子1的Npl值(0)，因此需要交换左右子树。那么最终得到的左偏堆还是符合推论

**Comment by XuebaStudy** [comment]
哎呀呀，我的问题，我忘了“\\"有转义，需要打两个,l
下面的图中，（）内为Npl，这棵树共10个节点，但右子树上有8个节点，明显大于log(10+1)（以2为底）
-----------0(1)--------
--------/------\\-------
----1(0)-------2(0)------
---------------/--------
-------------3(2)---------
-----------/------\\-------
------4(1)----------5(1)---
-----/----\\-------/----\\-----
---6(0)---7(0)--8(0)---9(0)---
我猜你的笔记应该是”最右路径“（一直向右走）而非”右子树“上有”至多有 ⌊log(N+1)⌋ 个节点“

**Reply by NoughtQ** [reply]
>我猜你的笔记应该是”最右路径“（一直向右走）而非”右子树“上有”至多有 ⌊log(N+1)⌋ 个节点“

对的对的，这里确实是我表述的问题，应该指最右侧路径，之后我会修改一下，感谢指出！

但我还有个疑问：为什么5是3的右孩子，不应该是2的右孩子吗？

**Reply by XuebaStudy** [reply]
如果5是2的右孩子，那我就应该把5放在第三层吧（与3同层），但事实上之前两种都是放在3的下面。总结而言，还是最后一种表示比较方便易懂😭
